tools/src/bin/docgen/main.rs

```diff
@@ -202,31 +202,33 @@ fn comment_and_requirement(
     src: &[u8],
 ) -> Result<(Option<Comment>, Option<Feature>), Box<dyn Error>> {
     let prev_prev = prev.prev_named_sibling();
-    let mut maybe_comment = Comment::new(prev, src).ok();
 
-    // If node wasn't a comment, see if it was an expression_statement
-    // that itself was preceded by a comment.  This skips over
-    // expression-like preprocessor attributes on function decls.
-    if let (None, "expression_statement", Some(prev_prev)) =
-        (&maybe_comment, prev.kind(), prev_prev)
-    {
-        maybe_comment = Comment::new(prev_prev, src).ok();
+    // In the simple case, `prev` is a comment and `prev_prev` may
+    // be a feature requirement.
+    if let Ok(comment) = Comment::new(prev, src) {
+        let feature = match prev_prev {
+            Some(prev_prev) => Feature::new(prev_prev, src).ok(),
+            None => None,
+        };
+        return Ok((Some(comment), feature));
     }
 
-    // If prev wasn't a comment, see if it was a feature requirement.
-    if maybe_comment.is_none() {
-        return Ok(match Feature::new(prev, src).ok() {
-            Some(feat_req) => (None, Some(feat_req)),
-            None => (None, None),
-        });
+    // If `prev` wasn't a comment, see if it was an expression_statement
+    // that itself was preceded by a comment.  This skips over
+    // expression-like preprocessor attributes on function decls.
+    if prev.kind() == "expression_statement" {
+        return match prev_prev {
+            Some(prev_prev) => comment_and_requirement(prev_prev, src),
```

Comment:
The introduction of recursion seems kind of sudden, maybe deserves its own commit?